File: src/test/java/cz/vhromada/catalog/service/MovieServiceTest.java
package cz.vhromada.catalog.service;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

import cz.vhromada.catalog.domain.Medium;
import cz.vhromada.catalog.domain.Movie;
import cz.vhromada.catalog.repository.MovieRepository;
import cz.vhromada.catalog.utils.MovieUtils;
import cz.vhromada.common.repository.MovableRepository;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.test.service.MovableServiceTest;

import org.junit.jupiter.api.Test;
import org.mockito.Mock;
import org.springframework.cache.Cache;

/**
 * A class represents test for class {@link MovieService}.
 *
 * @author Vladimir Hromada
 */
class MovieServiceTest extends MovableServiceTest<Movie> {

    /**
     * Instance of {@link MovieRepository}
     */
    @Mock
    private MovieRepository repository;

    /**
     * Test method for {@link MovieService#MovieService(MovieRepository, Cache)} with null repository for movies.
     */
    @Test
    void constructor_NullMovieRepository() {
        assertThatThrownBy(() -> new MovieService(null, getCache())).isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link MovieService#MovieService(MovieRepository, Cache)} with null cache.
     */
    @Test
    void constructor_NullCache() {
        assertThatThrownBy(() -> new MovieService(repository, null)).isInstanceOf(IllegalArgumentException.class);
    }

    @Override
    protected MovableRepository<Movie> getRepository() {
        return repository;
    }

    @Override
    protected MovableService<Movie> getService() {
        return new MovieService(repository, getCache());
    }

    @Override
    protected String getCacheKey() {
        return "movies";
    }

    @Override
    protected Movie getItem1() {
        return MovieUtils.newMovieDomain(1);
    }

    @Override
    protected Movie getItem2() {
        return MovieUtils.newMovieDomain(2);
    }

    @Override
    protected Movie getAddItem() {
        return MovieUtils.newMovieDomain(null);
    }

    @Override
    protected Movie getCopyItem() {
        final Movie movie = MovieUtils.newMovieDomain(1);
        movie.setId(null);
        for (final Medium medium : movie.getMedia()) {
            medium.setId(null);
        }

        return movie;
    }

    @Override
    protected Class<Movie> getItemClass() {
        return Movie.class;
    }

    @Override
    protected void assertDataDeepEquals(final Movie expected, final Movie actual) {
        MovieUtils.assertMovieDeepEquals(expected, actual);
    }

}

File: src/test/kotlin/cz/vhromada/catalog/service/MovieServiceTest.kt
package cz.vhromada.catalog.service

import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.verifyNoMoreInteractions
import com.nhaarman.mockitokotlin2.whenever
import cz.vhromada.catalog.domain.Movie
import cz.vhromada.catalog.repository.MovieRepository
import cz.vhromada.catalog.utils.MovieUtils
import cz.vhromada.common.provider.AccountProvider
import cz.vhromada.common.provider.TimeProvider
import cz.vhromada.common.service.MovableService
import cz.vhromada.common.test.service.MovableServiceTest
import cz.vhromada.common.test.utils.TestConstants
import org.mockito.Mock
import org.springframework.data.jpa.repository.JpaRepository

/**
 * A class represents test for class [MovieService].
 *
 * @author Vladimir Hromada
 */
class MovieServiceTest : MovableServiceTest<Movie>() {

    /**
     * Instance of [MovieRepository]
     */
    @Mock
    private lateinit var repository: MovieRepository

    /**
     * Instance of [AccountProvider]
     */
    @Mock
    private lateinit var accountProvider: AccountProvider

    /**
     * Instance of [TimeProvider]
     */
    @Mock
    private lateinit var timeProvider: TimeProvider

    override fun getRepository(): JpaRepository<Movie, Int> {
        return repository
    }

    override fun getAccountProvider(): AccountProvider {
        return accountProvider
    }

    override fun getTimeProvider(): TimeProvider {
        return timeProvider
    }

    override fun getService(): MovableService<Movie> {
        return MovieService(repository, accountProvider, timeProvider, cache)
    }

    override fun getCacheKey(): String {
        return "movies${TestConstants.ACCOUNT.id}"
    }

    override fun getItem1(): Movie {
        return MovieUtils.newMovieDomain(1)
    }

    override fun getItem2(): Movie {
        return MovieUtils.newMovieDomain(2)
    }

    override fun getAddItem(): Movie {
        return MovieUtils.newMovieDomain(null)
    }

    override fun getCopyItem(): Movie {
        val movie = MovieUtils.newMovieDomain(1)
                .copy(id = null)
        for (medium in movie.media) {
            medium.id = null
        }

        return movie
    }

    override fun initAllDataMock(data: List<Movie>) {
        whenever(repository.findByAuditCreatedUser(any())).thenReturn(data)
    }

    override fun verifyAllDataMock() {
        verify(repository).findByAuditCreatedUser(TestConstants.ACCOUNT_ID)
        verifyNoMoreInteractions(repository)
    }

    override fun anyItem(): Movie {
        return any()
    }

    override fun argumentCaptorItem(): KArgumentCaptor<Movie> {
        return argumentCaptor()
    }

    override fun assertDataDeepEquals(expected: Movie, actual: Movie) {
        MovieUtils.assertMovieDeepEquals(expected, actual)
    }

}
